socorro/signature/__main__.py

```diff
@@ -0,0 +1,229 @@
+Given crash ids, pulls down information from Socorro, generates signatures, and prints
+signature information.
+"""
+
+EPILOG = """
+Note: In order for the SignatureJitCategory rule to work, you need a valid API token from
+Socorro that has "View Personally Identifiable Information" permission.
+"""
+
+logger = logging.getLogger('socorro.signature')
+
+
+# FIXME(willkg): This hits production. We might want it configurable.
+API_URL = 'https://crash-stats.mozilla.com/api/'
+
+
+def setup_logging(logging_level):
+    dc = {
+        'version': 1,
+        'disable_existing_loggers': True,
+        'formatters': {
+            'bare': {
+                'format': '%(levelname)s: %(message)s'
+            }
+        },
+        'handlers': {
+            'console': {
+                'level': 'DEBUG',
+                'class': 'logging.StreamHandler',
+                'formatter': 'bare',
+            },
+        },
+        'loggers': {
+            'socorro': {
+                'propagate': False,
+                'handlers': ['console'],
+                'level': logging_level,
+            },
+        },
+    }
+    logging.config.dictConfig(dc)
+
+
+class OutputBase:
+    """Base class for outputter classes
+
+    Outputter classes are context managers. If they require start/top or begin/end semantics, they
+    should implement ``__enter__`` and ``__exit__``.
+
+    Otherwise they can just implement ``data`` and should be fine.
+
+    """
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        pass
+
+    def warning(self, line):
+        """Prints out a warning line to stderr
+
+        :arg str line: the line to print to stderr
+
+        """
+        print('WARNING: %s' % line, file=sys.stderr)
+
+    def data(self, crash_id, old_sig, new_sig, notes):
+        """Outputs a data point
+
+        :arg str crash_id: the crash id for the signature generated
+
+        :arg str old_sig: the old signature retrieved in the processed crash
+
+        :arg str new_sig: the new generated signature
+
+        :arg list notes: any processor notes
+
+        """
+        pass
+
+
+class TextOutput(OutputBase):
+    def data(self, crash_id, old_sig, new_sig, notes):
+        print('Original: %s' % old_sig)
+        print('New:      %s' % new_sig)
+        print('Same?:    %s' % (old_sig == new_sig))
+
+        if notes:
+            print('Notes:    (%d)' % len(notes))
+            for note in notes:
+                print('          %s' % notes)
+
+
+class CSVOutput(OutputBase):
+    def __enter__(self):
+        self.out = csv.writer(sys.stdout, quoting=csv.QUOTE_ALL)
+        self.out.writerow(['crashid', 'old', 'new', 'same?', 'notes'])
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.out = None
+
+    def data(self, crash_id, old_sig, new_sig, notes):
+        self.out.writerow([crash_id, old_sig, new_sig, str(old_sig == new_sig), notes])
```

Comment:
We could do a JSON output, but I'm going to wait until someone *needs* it. CSV is super helpful and convenient for regression testing.